packages/application-generic/src/modules/cron.module.ts

```diff
@@ -0,0 +1,99 @@
+import { DynamicModule, Module, Provider } from '@nestjs/common';
+import { JobCronNameEnum, JobTopicNameEnum } from '@novu/shared';
+import { DalService } from '@novu/dal';
+import {
+  ACTIVE_CRON_JOBS_TOKEN,
+  AgendaCronService,
+  CronService,
+  MetricsService,
+} from '../services';
+import { MetricsModule } from './metrics.module';
+import { dalService as customDalService } from '../custom-providers';
+import os from 'os';
+import { Agenda } from '@hokify/agenda';
+
+/**
+ * This map is a little uncomfortable because it depends on the Job topic name, coupling the Cron jobs to the
+ * queue names that are injected via the environment variable `ACTIVE_WORKERS`.
+ *
+ * Moving forward, we should consider specifying an enum for Workers to decouple the Worker names from
+ * the job names. This would allow us to specify the cron jobs in a more explicit way for each worker.
+ */
+const cronJobsFromWorkers: Partial<
+  Record<JobTopicNameEnum, Array<JobCronNameEnum>>
+> = {
+  [JobTopicNameEnum.STANDARD]: [
+    JobCronNameEnum.CREATE_BILLING_USAGE_RECORDS,
+    JobCronNameEnum.SEND_CRON_METRICS,
+  ],
+};
+
+export const cronService = {
+  provide: CronService,
+  useFactory: async (
+    metricsService: MetricsService,
+    activeCronJobs: JobCronNameEnum[],
+    dalService: DalService
+  ) => {
+    const agenda = new Agenda({
+      mongo: dalService.connection.getClient().db() as any,
+      /**
+       * Sets the hostname for the Job. Used to debug last host to run the job via the collection
+       *
+       * @see https://github.com/agenda/agenda/tree/master?tab=readme-ov-file#namename
+       */
+      name: os.hostname + '-' + process.pid,
+    });
+    const service = new AgendaCronService(
+      metricsService,
+      activeCronJobs,
+      agenda
+    );
+
+    return service;
+  },
+  inject: [MetricsService, ACTIVE_CRON_JOBS_TOKEN, DalService],
+};
+
+const PROVIDERS: Provider[] = [
+  cronService,
+  AgendaCronService,
+  MetricsService,
+  customDalService,
+];
+
+@Module({})
+export class CronModule {
+  static forRoot(activeWorkers: JobTopicNameEnum[]): DynamicModule {
+    return {
+      imports: [MetricsModule],
+      module: CronModule,
+      providers: [
+        {
+          provide: ACTIVE_CRON_JOBS_TOKEN,
```

Comment:
Custom provider to determine which Cron jobs to run, depending on the active worker. These "jobs to run" are dynamically injected into the CronService to conditionally add/run jobs.

Reply:
Great! I was actually thinking about suggesting something similar today so we could couple the initialization to the environment variables and not by code, but you've made it much more efficient! what do you think about extracting this provider to separate const in terms of readability, I think it could be under the "custom providers"


Just a quick note, in case you're not familiar with the current project initialization process: we currently spin up infrastructure-level instances based on the value of the ACTIVE_WORKERS environment variable. In the past, we had a couple of instances where ACTIVE_WORKERS was an empty array, which meant that they were initializing all of the Bull MQ workers. I'm not sure if we still have those instances, but if we do, just be aware that they will also initialize this Cron service.

Reply:
Great suggestion, I considered this however we do already have this as a "custom provider" (although not a Nest provider) in the Worker init config.

I think the action item here is to extract the [Worker init config logic](https://github.com/novuhq/novu/blob/2d0f1f6fd91e01c28bdccd5a4bc3bee6cbc70a56/apps/worker/src/config/worker-init.config.ts#L68) into an actual custom provider and have it live in a shared place, using an `ACTIVE_WORKERS` token as the provider reference.

I'm wary of refactoring the Worker config in this PR though because it will require additional testing effort. I'll raise it as an item for discussion at tech summit (and Worker queue/jobs application architecture more generally).

Reply:
See https://www.notion.so/novuhq/Tech-Summits-c9cf00a26717416f98152adbe6a6e88a?pvs=4#bebc0e92aa0b4bb0a90703d09f67f3cc